### PR TITLE
rune/skeleton/README: refresh the README.md

### PR DESCRIPTION
**Pull request by haosanzi**
Fixes: #292
Signed-off-by: Shirong Hao <shirong@linux.alibaba.com>

### CODE REVIEW

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 1]
```diff
@@ -4,19 +4,75 @@ Note that this code base is inspired by [v28 SGX in-tree driver](https://patchwo
...
 ---
 
-# Run skeleton with Docker
-## Build liberpal-skeleton
+# Introduction
 
-`skeleton` depends on `protobuf-c` compiler. Please refer to [this step](https://github.com/protobuf-c/protobuf-c#building) to install it on your platform.
+Skeleton is an example of enclave runtime, interfacing with Enclave Runtime PAL API for easy interfacing with `rune`.
+
+Based on many engineering practices, skeleton is proven as a best practice on experimental features and PoC. Therefore, some preview features will be implemented in skeleton firstly.
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:11]: s/best practice on experimental features/best practice on implementing experimental features/

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 2]
```diff
@@ -4,19 +4,75 @@ Note that this code base is inspired by [v28 SGX in-tree driver](https://patchwo
...
 
 ```shell
 cd "${path_to_inclavare_containers}/rune/libenclave/internal/runtime/pal/skeleton"
 make
 cp liberpal-skeleton-v*.so /usr/lib
 ```
-Debug enclave is generated by default. Please use `make PRODUCT_ENCLAVE=1` command to generate production enclave.
 
 ## Build skeleton container image
+
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:73]: Rename all 'container image' as 'docker image'.

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 3]
```diff
@@ -122,28 +187,84 @@ where:
...
+| Options                             | Function                                                                                               |
+| ----------------------------------- | -------------------------------------------------------------------------------------------------------|
+| -p,  --product                      | Generate production enclave. Debug enclave is generated by default.                                    |
+| -s, --mmap-size value               | Launch an enclave with memory size equals to value(bytes).                                             |
+| -a,  --attrs value                  | Launch an enclave with attributes equals to value(hex).                                                |
+| -x, --xfrm value                    | Launch an enclave with xfrm equals to value(hex).                                                      |
+| -n, --null\_dereference\_protection | Enable [Enclave NULL dereference protection](#enclave-null-dereference-protection). Disable by default.|
+| -m,  --mman\_min\_addr value        | Launch an enclave with mman\_min\_addr equals to value(hex).                                           |
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:228]: Keep the same numbers of space between short and long options.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 4]
```diff
@@ -169,25 +290,8 @@ cd "$HOME/rune_workdir/rune-container"
...
-In order to enable this protection for OOT driver, the restriction from
-mmapping must be disabled:
-```shell
-sudo sysctl -w vm.mmap_min_addr=0
-```
-in-tree driver doesn't have to do it.
-
-FIXME: Current implementation assuems the build, signing and running stages
-are on the same platform.
+Please refer to this guide to run skeleton with kvmtool.
```
Comment (hustliyilin) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:297]: where is the URL of `the guide`?

Comment (haosanzi) [reply rune/libenclave/internal/runtime/pal/skeleton/README.md:297]: The guide is not ready. @uudiin will provide the guide in the next release. #419

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 5]
```diff
@@ -122,28 +187,84 @@ where:
 - @enclave.runtime.path: specify the path to enclave runtime to launch.
 - @enclave.runtime.args: specify the specific arguments to enclave runtime, seperated by the comma.
 
-## Run skeleton
+### Run skeleton
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:190]: Why use ### not ##?

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 6]
```diff
@@ -122,28 +187,84 @@ where:
...
+
 Assuming you have an OCI bundle from the previous step you can execute the container in this way.
 
 ```shell
 cd "$HOME/rune_workdir/rune-container"
 sudo rune run skeleton-enclave-container
 ```
 
-## Run skeleton with rune attest command
+### Run skeleton with rune attest command
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:199]: Move attest function to features.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Review by jiazhang0 — APPROVED** [review]

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 7]
```diff
@@ -122,28 +187,85 @@ where:
...
+read the zero page fed with malicious data, or write the confidential data
+to zero page.
+
+In order to prevent from this attack, skeleton enclave runtime implements
+enclave NULL dereference protection for OOT and in-tree drivers. However,
+OOT and in-tree drivers have different designs and implementations. This
+also affects the behavior of enclave NULL dereference protection.
+
+In order to enable this protection for OOT driver, the restriction from
+mmapping must be disabled:
```
Comment (hustliyilin) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:246]: add an empty line here please.

File: rune/libenclave/internal/runtime/pal/skeleton/README.md [hunk 8]
```diff
@@ -169,25 +291,8 @@ cd "$HOME/rune_workdir/rune-container"
...
 
----
-
-# Enclave NULL dereference protection
+## Enclave VM
 
-The potential attacker may re-map zero page to induce the buggy enclave to
-read the zero page fed with malicious data, or write the confidential data
-to zero page.
+Skeleton now can interact with kvmtool to support enclave VM for demonstration. This is the first step to support a new form of enclave.
```
Comment (hustliyilin) [inline rune/libenclave/internal/runtime/pal/skeleton/README.md:296]: `kvmtool` should be given a URL when When it first appeared in the README.nd

**Review by hustliyilin — APPROVED** [review]

**Review by YangLiang3 — APPROVED** [review]

**Pull request merged**